src/type_checker/checker.rs

```diff
@@ -258,7 +258,7 @@ impl<B: Backend> TypeChecker<B> {
                 // and is of the same type as the rhs
                 let rhs_typ = self.compute_type(rhs, typed_fn_env)?.unwrap();
 
-                if !rhs_typ.typ.match_expected(&lhs_node.typ) {
+                if !rhs_typ.typ.match_expected(&lhs_node.typ, true) {
```

Comment:
This type check works because when it comes to assignment expression the assumption is that the variable is under mutable mode.
So it should allow assigning either constant or non constant field to the variable.